app/components/download-switch/download-switch.tsx

```diff
@@ -167,11 +167,15 @@ export function DownloadSwitch({
 
   const handleDownload = async (
     track: Track,
-    download: () => Promise<FileSystemDownloadResult>,
+    download: () => Promise<FileSystemDownloadResult | undefined>,
   ) => {
     // TODO Check `newAbortSignal.aborted` after every `await`?
     statuses.set(track.trackId, DownloadStatus.Downloading)
-    const { uri, md5, status: statusCode } = await download()
+    const downloadResult = await download()
+    if (downloadResult === undefined) {
+      return
```

Comment:
Not sure if early returning here is correct. Especially since `statuses.set` has already been set. Should I move the early return and `download()` call to the top of this function?

Reply:
With the current implementation and status logic, the status `Downloading` needs to be set before `await download`. I would consider `downloadResult` to be `undefined` to be a rejected promise and throw an exception instead o returning. This is a general problem: Whenever something fails in this function before we set the status to `Downloaded` we have the state `Downloading` although nothing is happening anymore. In a proper implementation, when the downloading fails for a reason that does not stem from the user stopping it by pressing a button, we should clean-up and in particular set the status to `NotDownloaded`. However, if the download fails because the user stopped it on purpose, then that stopping action already set the status to something else and we must not overwrite that status. If you come up with a good design that encompasses those cases in a good way like some proper state machine implementation, then go for it. I'm currently too confused (maybe after the vacation ...).

Reply:
Oh boy. I'm confused as well, but at least understand the issue now.

I'll change the early return to throwing an error to have the same behavior as the previous implementation (if `downloadResult` would've been `undefined` the code would've thrown an error due to `const { uri, md5, status: statusCode } = downloadResult`).

I'll think about proper error handling a bit, need to fully understand the control flow first, though.

Reply:
Error thrown in 88852db.